### PR TITLE
Ch cache nutritional info 142465437

### PR DESCRIPTION
**Pull request by Andela-Didacus**
**what does this pull request do?**
adds a cache for the nutritional info and add test

### CODE REVIEW

File: wger/nutrition/models.py [hunk 1]
```diff
@@ -624,14 +637,18 @@ def get_nutritional_values(self, use_metric=True):
...
+        if not nutritional_info:
+            nutritional_info = {'energy': 0,
+                                'protein': 0,
+                                'carbohydrates': 0,
+                                'carbohydrates_sugar': 0,
+                                'fat': 0,
+                                'fat_saturated': 0,
+                                'fibres': 0,
+                                'sodium': 0}
+            cache.set("nutritional_info", nutritional_info)
```
Comment (andela-osule) [inline wger/nutrition/models.py]: Let move this to the line before returning `nutritional_info`

File: wger/nutrition/tests/test_cache.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from django.core.cache import cache
+from django.test import TestCase
+from django.test.utils import override_settings
+from wger.nutrition.models import NutritionPlan, Meal, MealItem
+
+
+class MyTestCase(TestCase):
+    def test_caching(self):
```
Comment (andela-osule) [inline wger/nutrition/tests/test_cache.py]: This test can do better stressing out `Meal.get_nutritional_values`. Please compare the time elapsed calling that method twice. 
```
start_time = time.time()
# call 1
elapsed_time_1 = time.time() - start_time
start_time = time.time()
# call 2
elapsed_time_2 = time.time() - start_time
self.assertGreater(elapsed_time_2, elapsed_time_1)
```

File: wger/nutrition/models.py [hunk 3]
```diff
@@ -546,27 +548,44 @@ def get_nutritional_values(self, use_metric=True):
...
+                values = item.get_nutritional_values(use_metric=use_metric)
+                for key in nutritional_info.keys():
+                    nutritional_info[key] += values[key]
+
+            # Only 2 decimal places, anything else doesn't make sense
+            for i in nutritional_info:
+                nutritional_info[i] = Decimal(nutritional_info[i]).quantize(TWOPLACES)
+
+            cache.set("nutritional_info_meal", nutritional_info)
+        return nutritional_info
```
Comment (andela-osule) [inline wger/nutrition/models.py]: Use the corresponding decorator to connect to this handler. See django's documentation

File: wger/nutrition/models.py [hunk 4]
```diff
@@ -546,27 +548,44 @@ def get_nutritional_values(self, use_metric=True):
...
+            cache.set(self.get_nutritional_values(self), self, nutritional_info)
 
-        # Only 2 decimal places, anything else doesn't make sense
-        for i in nutritional_info:
-            nutritional_info[i] = Decimal(nutritional_info[i]).quantize(TWOPLACES)
+    def save(self, *args, **kwargs):
+        '''
+        Reset the cache
+        '''
+        pre_save.connect(self.pre_save_cache)
```
Comment (andela-osule) [inline wger/nutrition/models.py]: We don't want to connect to the callback this way. Defeats the purpose for using signals which is to decouple event handling from the model.

File: wger/nutrition/models.py [hunk 5]
```diff
@@ -624,54 +643,71 @@ def get_nutritional_values(self, use_metric=True):
...
+            # Only 2 decimal places, anything else doesn't make sense
+            for i in nutritional_info:
+                nutritional_info[i] = Decimal(nutritional_info[i]).quantize(TWOPLACES)
+            cache.set("nutritional_info_mealitem", nutritional_info)
+        return nutritional_info
 
-                # Energy is not a weight!
-                if key == 'energy':
-                    continue
+    def pre_save_cache(sender, instance, **kwargs):
```
Comment (andela-osule) [inline wger/nutrition/models.py]: These handlers should be functions and not instance methods. Please see the documentation

File: wger/nutrition/models.py [hunk 6]
```diff
@@ -624,54 +643,71 @@ def get_nutritional_values(self, use_metric=True):
...
+        cache.delete(self.get_nutritional_values(self), self, nutritional_info)
 
-                # Everything else, to ounces
-                nutritional_info[key] = AbstractWeight(value, 'g').oz
+    def post_save_cache(sender, instance, **kwargs):
```
Comment (andela-osule) [inline wger/nutrition/models.py]: Same as here. See the documentation about connecting handlers using decorators.

File: wger/nutrition/models.py [hunk 7]
```diff
@@ -624,54 +639,67 @@ def get_nutritional_values(self, use_metric=True):
...
+            cache.set("nutritional_info_mealitem", nutritional_info)
+        return nutritional_info
 
-                # Energy is not a weight!
-                if key == 'energy':
-                    continue
 
-                # Everything else, to ounces
-                nutritional_info[key] = AbstractWeight(value, 'g').oz
+@receiver(pre_save, sender=MealItem)
```
Comment (andela-osule) [inline wger/nutrition/models.py]: We're should either use a pre_save or a post_save to delete the cache.

File: wger/nutrition/models.py [hunk 8]
```diff
@@ -624,54 +639,67 @@ def get_nutritional_values(self, use_metric=True):
...
+def pre_save_cache(sender, instance, **kwargs):
+    cache.delete(self.get_nutritional_values(self), self, nutritional_info)
 
-        # Only 2 decimal places, anything else doesn't make sense
-        for i in nutritional_info:
-            nutritional_info[i] = Decimal(nutritional_info[i]).quantize(TWOPLACES)
 
-        return nutritional_info
+@receiver(post_save, sender=MealItem)
```
Comment (andela-osule) [inline wger/nutrition/models.py]: Please explain why we're doing this

File: wger/nutrition/models.py [hunk 9]
```diff
@@ -546,27 +548,40 @@ def get_nutritional_values(self, use_metric=True):
...
+@receiver(pre_save, sender=Meal)
+def pre_save_cache(sender, instance, **kwargs):
+    cache.delete(self.get_nutritional_values(self), self, nutritional_info)
 
-        # Only 2 decimal places, anything else doesn't make sense
-        for i in nutritional_info:
-            nutritional_info[i] = Decimal(nutritional_info[i]).quantize(TWOPLACES)
 
-        return nutritional_info
+@receiver(post_save, sender=Meal)
```
Comment (andela-osule) [inline wger/nutrition/models.py]: Why are we doing this?

File: wger/nutrition/models.py (hunk 8 above)
Comment (daisyndungu) [inline wger/nutrition/models.py]: This saves the cache after saving..

File: wger/nutrition/models.py [hunk 10]
```diff
@@ -546,27 +548,35 @@ def get_nutritional_values(self, use_metric=True):
...
+            cache.set("nutritional_info_meal", nutritional_info)
+        return nutritional_info
 
-        # Only 2 decimal places, anything else doesn't make sense
-        for i in nutritional_info:
-            nutritional_info[i] = Decimal(nutritional_info[i]).quantize(TWOPLACES)
 
-        return nutritional_info
+@receiver(pre_save, sender=Meal)
+def pre_save_cache(sender, instance, **kwargs):
```
Comment (andela-osule) [inline wger/nutrition/models.py]: Can we name this function according to what it does? A name like uncache_nutrition_info_meal is more descriptive that what we have here. Also, this handler is overwritten down where a handler gets defined for presave signal on MealItem

File: wger/nutrition/tests/test_cache.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+from wger.core.tests.base_testcase import WorkoutManagerTestCase
+
+
+class MyTestCase(WorkoutManagerTestCase):
+    def test_caching(self):
+        # clear cache
+        cache.clear()
+
+        Meal.get_nutritional_values(self)
```
Comment (andela-osule) [inline wger/nutrition/tests/test_cache.py]: We shouldn't call instance methods this way.

```
meal = Meal.objects.earliest('pk')

start_time = time.time()
meal.get_nutritional_values()
elapsed_time_before = time.time() - start_time
```

File: wger/nutrition/tests/test_cache.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        start_time = time.time()
+        elapsed_time_before = time.time() - start_time
+        start_time = time.time()
+        Meal.get_nutritional_values(self)
```
Comment (andela-osule) [inline wger/nutrition/tests/test_cache.py]: :(

File: wger/nutrition/tests/test_cache.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        # taken to load from cache
+        elapsed_time_2 = time.time() - start_time
```
Comment (andela-osule) [inline wger/nutrition/tests/test_cache.py]: This can be better named `elapsed_time_after` like `elapsed_time_before` up there.

**Pull request merged**